designs/feature-control.md

```diff
@@ -0,0 +1,85 @@
+Feature Access Control for Unreleased Features
+====================================
+
+What is the problem?
+--------------------
+
+Currently, we cannot push large WIP features to the `master` branch as new builds will include unfinished code. Additionally, development in separate, isolated feature branches cause massive merges resulting in days of work to remedy. A feature access control system will let us develop code and push it to the `master` branch as it's written, while hiding the functionality from the general populace.
+
+What will be added?
+---------------------
+
+We will add a feature access control class to the codebase which will allow developers to gate off code paths and allow easily-configurable access to users.
+
+### Success criteria for the addition
+
+* A feature access control system that grants access to unreleased code for development purposes.
+  * Access should not be dependent on a rebuild--this should be user-configurable.
+  * Flags should be undocumented outside the codebase. Even though features will be included in new builds, they will eminently not be ready for general use.
+  * Individual features should be able to be toggled on/off on-the-fly or only on restart ("session-permanent") in a predictable manner.
+  * Tests should be able to inject their own configurations into the feature access control system to ensure correct features are tested.
+
+### Out-of-Scope
+
+* Visual confirmation for a user to discern whether or not a feature is session-permanent when adding a toggle to their `settings.json` file while the extension is running.
+* The ability to pass non-boolean values through feature access control.
+* The ability to permanently override the switch for the feature (as opposed to pruning the toggle and the forking code) when a feature is set to go live.
```

Comment:
I think we should be more opinionated about the lifecycle of these toggles - what's the removal strategy?

You may actually want the ability to switch a toggle permanently _before_ removal, this is especially useful for experimental bug-fixes; a typical flow might be:

1. Customer raises an issue on Github - specific to their setup (e.g. proxy issue)
2. Gated bug-fix introduced, customer informed and asked to provide feedback to verify bug is fixed.
3. Once customer confirms fix resolves issue, the default is switched to 'on for all'
4. After some period of time (?) the toggle and old code is removed

The advantage of this approach is it gives us an ability to easily roll-back a fix if it actually causes more wide-spread problems.

Reply:
My opinion on this is that we don't want orphaned code paths, which I feel like a permanent setting would open the way for...as good as our intentions are, we might forget about these going forward.

Essentially, this is the implementation of my opinion: by disallowing the option in the first place, it'll force developers to have to remove the branching path in order to make the feature generally accessible.

I'm open to debate on this.

Reply:
Absolutely agree we don't want orphan code-paths, which is why I'm suggesting we have a defined lifecycle for how long a toggle should live. It should be removed aggressively - you just might want an ability to 'turn it on for a bit'.

All feature toggles should be immediately marked as `deprecated` and we should be actively resolving all deprecations.

Reply:
I think every feature flag added should have a corresponding "Remove this feature flag" issue filed at the same time. 

* If a major feature is gated while under development, its inaugural release would see the gate removed, and the issue would be fulfilled
* If a community support bug-fix is gated (and released) for trial and feedback, once we have confirmation that the fix is valid, the gate can be removed within a reasonable timeframe, fulfilling the issue. A release should be made sooner rather than later, to avoid requesting more users to enable the feature flag instead of waiting for the next release.

Reply:
I think in code there should be a target removal date (or release version) as part of the actual implementation (that is then linted for as part of build). This is changeable of course but would have to be actively changed.

Reply:
Should this linting break the build if it's in violation, or should this just be a strongly-worded warning? Should this affect any users using an out-of-date feature (either preventing them from using the feature or popping up a strongly-worded warning)?

The worst case scenario with a build-blocker is that this could affect other developers--if a contributor is working on a fix but can't build due to an expired feature, should they be expected to resolve the issue themselves or sit on their hands until the owner/one of us changes the feature?